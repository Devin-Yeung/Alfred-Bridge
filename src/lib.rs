use serde::Serialize;
use typed_builder::TypedBuilder;

pub mod item;

#[derive(Serialize, TypedBuilder)]
#[builder(
    doc,
    field_defaults(default),
    mutators(
        fn item(&mut self, item: item::AlfredItem) -> &mut Self {
            self.items.push(item);
            self
        }
    )
)]
pub struct AlfredOutput {
    #[builder(via_mutators)]
    items: Vec<item::AlfredItem>,
}

impl AlfredOutput {
    pub fn stdout(&self) -> Result<(), serde_json::Error> {
        println!("{}", serde_json::to_string(&self)?);
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_output() {
        let output = AlfredOutput::builder()
            .item(
                item::AlfredItem::builder()
                    .title("Hello, World!")
                    .subtitle("This is a test subtitle.")
                    .arg("https://example.com")
                    .icon(
                        item::Icon::builder()
                            .r#type(item::IconType::FileIcon)
                            .path("icon.png")
                            .build(),
                    )
                    .build(),
            )
            .build();

        output.stdout().unwrap(); // This will print the JSON to stdout
        insta::assert_json_snapshot!(output)
    }
}
