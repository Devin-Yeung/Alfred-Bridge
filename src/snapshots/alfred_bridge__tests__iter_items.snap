---
source: src/lib.rs
expression: output
---
{
  "items": [
    {
      "title": "Item 1",
      "valid": true,
      "type": "default"
    },
    {
      "title": "Item 2",
      "valid": true,
      "type": "default"
    }
  ]
}
